src/transformStateWithClones.js

```diff
@@ -7,7 +7,31 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const newState = { ...state };
+  const stateResult = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(newState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete newState[key];
+        }
+        break;
+
+      case 'clear':
+        for (const object in newState) {
+          delete newState[object];
+        }
+        break;
+    }
+    stateResult.push({ ...newState });
```

Comment:
Cloning `newState` before pushing it to `stateResult` ensures that each state in the result is independent of subsequent changes. This is a good practice.